File: lib/ppurio.ts
// 뿌리오 SMS 발송 API

const PPURIO_API_BASE = 'https://message.ppurio.com/v1'
const PPURIO_ACCOUNT = process.env.PPURIO_ACCOUNT!
const PPURIO_ACCESS_KEY = process.env.PPURIO_ACCESS_KEY!
const PPURIO_FROM_NUMBER = process.env.PPURIO_FROM_NUMBER || '1866-1845'

// 토큰 캐시 (24시간 유효)
let cachedToken: string | null = null
let tokenExpiry: number = 0

/**
 * 뿌리오 Access Token 발급 (24시간 유효)
 */
async function getAccessToken(): Promise<string> {
  // 캐시된 토큰이 유효하면 재사용
  if (cachedToken && Date.now() < tokenExpiry) {
    return cachedToken
  }

  const tokenUrl = `${PPURIO_API_BASE}/token`
  const credentials = Buffer.from(`${PPURIO_ACCOUNT}:${PPURIO_ACCESS_KEY}`).toString('base64')

  try {
    const response = await fetch(tokenUrl, {
      method: 'POST',
      headers: {
        'Authorization': `Basic ${credentials}`,
        'Content-Type': 'application/json'
      }
    })

    if (!response.ok) {
      throw new Error(`뿌리오 토큰 발급 실패: ${response.status}`)
    }

    const data = await response.json()

    if (!data.token) {
      throw new Error('토큰 응답에 token 필드 없음')
    }

    // 토큰 캐시 (23시간 유효로 설정)
    cachedToken = data.token
    tokenExpiry = Date.now() + (23 * 60 * 60 * 1000)

    console.log('✅ 뿌리오 토큰 발급 성공')
    return data.token
  } catch (error) {
    console.error('❌ 뿌리오 토큰 발급 오류:', error)
    throw error
  }
}

/**
 * SMS 발송 인터페이스
 */
export interface SMSParams {
  to: string | string[] // 수신 번호 (단일 또는 배열)
  from?: string // 발신 번호 (선택, 기본값 사용)
  content: string // 문자 내용
  subject?: string // LMS 제목 (선택)
  refKey?: string // 참조 키 (선택)
}

/**
 * SMS 발송 (단문/장문 자동 감지)
 */
export async function sendSMS(params: SMSParams): Promise<boolean> {
  try {
    // 토큰 발급
    const token = await getAccessToken()

    // 수신자 번호 배열로 변환
    const recipients = Array.isArray(params.to) ? params.to : [params.to]

    // 문자 타입 자동 감지 (SMS: 90자 이하, LMS: 91자 이상)
    const messageType = params.content.length <= 90 ? 'SMS' : 'LMS'

    // 수신자 목록 구성
    const targets = recipients.map(phoneNumber => ({
      to: phoneNumber.replace(/-/g, ''), // 하이픈 제거
      name: '', // 이름은 선택사항
    }))

    // 발송 데이터 구성
    const sendData = {
      account: PPURIO_ACCOUNT,
      messageType: messageType,
      from: (params.from || PPURIO_FROM_NUMBER).replace(/-/g, ''),
      content: params.content,
      subject: params.subject || '',
      duplicateFlag: 'Y', // 중복 번호 허용
      targetCount: targets.length,
      targets: targets,
      refKey: params.refKey || `sms_${Date.now()}`
    }

    console.log('📤 SMS 발송 요청:', {
      messageType,
      targetCount: targets.length,
      contentLength: params.content.length
    })

    // API 호출
    const response = await fetch(`${PPURIO_API_BASE}/message`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(sendData)
    })

    if (!response.ok) {
      const errorText = await response.text()
      console.error('❌ SMS 발송 실패:', response.status, errorText)
      throw new Error(`SMS 발송 실패: ${response.status}`)
    }

    const result = await response.json()
    console.log('✅ SMS 발송 성공:', result)

    return true
  } catch (error) {
    console.error('❌ SMS 발송 오류:', error)
    // SMS 발송 실패는 견적 신청 자체를 막지 않음 (silent fail)
    return false
  }
}

/**
 * 견적 신청 알림 SMS 발송
 */
export async function sendConsultationNotification(data: {
  name: string
  phone: string
  address: string
  cameraCount: string
  referrer?: string
}) {
  // 직원 전화번호 목록 (환경변수)
  const staffNumbers = process.env.PPURIO_STAFF_NUMBERS?.split(',') || []

  if (staffNumbers.length === 0) {
    console.warn('⚠️ 직원 전화번호 목록이 설정되지 않았습니다 (PPURIO_STAFF_NUMBERS)')
    return false
  }

  // SMS 내용 구성
  const content = `[CCTV 견적 문의]
고객명: ${data.name}
설치 대수: ${data.cameraCount}
지역: ${data.address}
전화번호: ${data.phone}

유입 링크: ${data.referrer || '직접 접속'}`

  // 직원들에게 SMS 발송
  return await sendSMS({
    to: staffNumbers,
    content: content,
    refKey: `consultation_${Date.now()}`
  })
}

File: sms-server/ppurio.py
"""
뿌리오 SMS 발송 API (공식 예제 코드 기반)
거의 원본 그대로 사용하여 오류 최소화
"""

import logging
import requests
from requests.auth import HTTPBasicAuth
from typing import List, Optional
import os

API_URL = "https://message.ppurio.com"

class PpurioClient:
    def __init__(self, username: str, token: str):
        self.username = username
        self.token = token
        self.access_token = None

    def make_request(self, url, auth=None, payload=None, headers=None):
        """뿌리오 공식 예제 코드의 make_request 함수 (거의 그대로)"""
        response = None
        try:
            response = requests.post(url, headers=headers, json=payload, auth=auth, timeout=5)
            response.raise_for_status()
        except requests.exceptions.RequestException as e:
            if response:
                logging.error(f"An error occurred: {response.json()}")
            else:
                logging.error(f"Request failed: {e}")
            raise

        return response.json()

    def get_access_token(self):
        """엑세스 토큰 발급 (한 번 발급된 토큰은 24시간동안 유효합니다.)"""
        # 뿌리오 공식 예제 코드 그대로
        url = f"{API_URL}/v1/token"
        auth = HTTPBasicAuth(self.username, self.token)

        # response 데이터 생성
        response_data = self.make_request(url, auth)

        # access_token 반환
        token = response_data.get("token") if response_data else None

        if token:
            self.access_token = token
            logging.info("✅ 뿌리오 토큰 발급 성공")

        return token

    def send_sms(
        self,
        to_numbers: List[str],
        content: str,
        from_number: str,
        ref_key: str = "sms_notification"
    ) -> bool:
        """
        SMS 발송 (뿌리오 공식 예제 기반)

        Args:
            to_numbers: 수신자 번호 리스트
            content: 메시지 내용
            from_number: 발신 번호
            ref_key: 참조 키
        """
        # 토큰이 없으면 발급
        if not self.access_token:
            self.get_access_token()

        # 메시지 타입 자동 결정 (SMS: 90자 이하, LMS: 91자 이상)
        message_type = "SMS" if len(content.encode('utf-8')) <= 90 else "LMS"

        # request 데이터 세팅 (뿌리오 공식 예제 구조 그대로)
        url = f"{API_URL}/v1/message"
        headers = {
            "Authorization": f"Bearer {self.access_token}",
            "Content-Type": "application/json",
        }

        # targets 리스트 구성
        targets = [
            {
                "to": number.replace("-", ""),  # 하이픈 제거
                "name": ""
            }
            for number in to_numbers
        ]

        payload = {
            "account": self.username,
            "messageType": message_type,
            "content": content,
            "from": from_number.replace("-", ""),
            "duplicateFlag": "Y",  # 중복 허용
            "refKey": ref_key,
            "targetCount": len(targets),
            "targets": targets,
        }

        logging.info(f"📤 SMS 발송 요청: {message_type}, {len(targets)}명")

        try:
            # response 데이터 생성
            response_data = self.make_request(url, None, payload, headers)
            logging.info(f"✅ SMS 발송 성공: {response_data}")

            # messageKey 반환
            message_key = response_data.get("messageKey") if response_data else None
            return message_key is not None

        except Exception as e:
            logging.error(f"❌ SMS 발송 실패: {e}")
            return False
